File: frontend/src/types/books.ts
/**
 * 가계부 관련 타입 정의
 * 백엔드 models/books.py와 동기화
 */

export enum BookRole {
  OWNER = 'owner',
  EDITOR = 'editor',
}

export interface Book {
  id: string;
  owner_id: string;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface BookListItem extends Book {
  current_role: BookRole;
}

export interface BookCreate {
  name: string;
}

export interface BookUpdate {
  name: string;
}

export interface BookMember {
  book_id: string;
  user_id: string;
  email: string;
  full_name: string | null;
  role: BookRole;
  joined_at: string;
}

export interface BookMemberInvite {
  email: string;
  role: BookRole;
}

export interface BookMemberUpdate {
  role: BookRole;
}

File: backend/app/models/books.py
"""가계부 및 멤버 관련 Pydantic 모델."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field


class BookRole(str, Enum):
    """가계부 멤버 역할."""

    OWNER = "owner"
    EDITOR = "editor"


class Book(BaseModel):
    """가계부 단일 항목."""

    id: UUID
    owner_id: UUID
    name: str
    created_at: datetime
    updated_at: datetime


class BookListItem(Book):
    """목록 조회 응답."""

    current_role: BookRole


class BookCreate(BaseModel):
    """가계부 생성 요청."""

    name: str = Field(min_length=1, max_length=80)


class BookUpdate(BaseModel):
    """가계부 수정 요청."""

    name: str = Field(min_length=1, max_length=80)


class BookMember(BaseModel):
    """가계부 멤버 정보."""

    book_id: UUID
    user_id: UUID
    email: EmailStr
    full_name: str | None
    role: BookRole
    joined_at: datetime


class BookMemberInvite(BaseModel):
    """멤버 초대 요청."""

    email: EmailStr
    role: BookRole = BookRole.EDITOR


class BookMemberUpdate(BaseModel):
    """멤버 역할 수정 요청."""

    role: BookRole
